# EXAM 2 REVIEW

## Example 1: `mtcars`

* Load the dataset and show the dimensions of the dataset

In [3]:
library(MASS)
mtcars
dim(mtcars)

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Mazda RX4           21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag       21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Datsun 710          22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive      21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout   18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2   
Valiant             18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1   
Duster 360          14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
Merc 240D           24.4 4   146.7  62 3.69 3.190 20.00 1  0  4    2   
Merc 230            22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2   
Merc 280            19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4   
Merc 280C           17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4   
Merc 450SE          16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3   
Merc 450SL          17.3 8   275.8 180 3.07 3.730 17.60 0  0  3    3   
Merc 450SLC         15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3   
Cadillac Fleetwood  10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4   
Lincoln Continental 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4   
Chrysler Imperial   14.7 8   440.0 230 3.23 5.345 17.42 0  0  3    4   
Fiat 128            32.4 4    78.7  66 4.08 2.200 19.47 1  1  4    1   
Honda Civic         30.4 4    75.7  52 4.93 1.615 18.52 1  1  4    2   
Toyota Corolla      33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1   
Toyota Corona       21.5 4   120.1  97 3.70 2.465 20.01 1  0  3    1   
Dodge Challenger    15.5 8   318.0 150 2.76 3.520 16.87 0  0  3    2   
AMC Javelin         15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2   
Camaro Z28          13.3 8   350.0 245 3.73 3.840 15.41 0  0  3    4   
Pontiac Firebird    19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2   
Fiat X1-9           27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1   
Porsche 914-2       26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2   
Lotus Europa        30.4 4    95.1 113 3.77 1.513 16.90 1  1  5    2   
Ford Pantera L      15.8 8   351.0 264 4.22 3.170 14.50 0  1  5    4   
Ferrari Dino        19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6   
Maserati Bora       15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8   
Volvo 142E          21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2

[1] 32 11

In [6]:
names(mtcars)

[1] "mpg"  "cyl"  "disp" "hp"   "drat" "wt"   "qsec" "vs"   "am"   "gear"
[11] "carb"

* Fit a MLR and estimate the parameters
  

In [8]:
cars_reg = lm(mpg ~ ., data=mtcars)
summary(cars_reg)


Call:
lm(formula = mpg ~ ., data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.4506 -1.6044 -0.1196  1.2193  4.6271 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)  
(Intercept) 12.30337   18.71788   0.657   0.5181  
cyl         -0.11144    1.04502  -0.107   0.9161  
disp         0.01334    0.01786   0.747   0.4635  
hp          -0.02148    0.02177  -0.987   0.3350  
drat         0.78711    1.63537   0.481   0.6353  
wt          -3.71530    1.89441  -1.961   0.0633 .
qsec         0.82104    0.73084   1.123   0.2739  
vs           0.31776    2.10451   0.151   0.8814  
am           2.52023    2.05665   1.225   0.2340  
gear         0.65541    1.49326   0.439   0.6652  
carb        -0.19942    0.82875  -0.241   0.8122  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.65 on 21 degrees of freedom
Multiple R-squared:  0.869,	Adjusted R-squared:  0.8066 
F-statistic: 13.93 on 10 and 21 DF,  p-value: 3.793e-07

* $H_0$: hp is 0 VS $H_a$: hp is not 0 with $\alpha=0.05$

Since $p=0.3350>\alpha$, we fail to reject the null hypothesis. Therefore, we can conclude that hp is not actually in the model

* $H_0$: $\beta_1=\beta_2=...=\beta_11=0$ i.e. is there a linear relationship?

THE ANSWER FROM THIS COMES FROM THE F-STATISTIC AT THE BOTTOM OF THE SUMMARY OUTPUT.

This p-value is incredibly small, so we can conclude that a linear relationship does indeed exist

* Interpret the regression coefficient of horsepower

When the rest of the variables are held constant, if you increase horsepower by one unit, you decrease the mpg by 0.02

* $H_0$: hp = am = wt = 0 VS $H_a$: one of them not equal to zero 

NEED TO BUILD A REDUCED MODEL, it includes every variable EXCEPT those variables that we want in our hypothesis test

THEN we need to stick it into an anova table

In [15]:
names(mtcars)

[1] "mpg"  "cyl"  "disp" "hp"   "drat" "wt"   "qsec" "vs"   "am"   "gear"
[11] "carb"

In [16]:
car_red_reg = lm(mpg ~ cyl + disp + drat + qsec + vs + gear + carb, data=mtcars)
anova(car_red_reg, cars_reg)

Res.Df RSS      Df Sum of Sq F        Pr(>F)   
1 24     189.8998 NA       NA        NA        NA
2 21     147.4944  3 42.40536  2.012534 0.1430042

* Interpret $R^2$

Since $R^2=0.869$, so 86.9% of our variability can be explained by our model

* Calculate the correlation matrix

In [20]:
cor(mtcars)

mpg        cyl        disp       hp         drat        wt        
mpg   1.0000000 -0.8521620 -0.8475514 -0.7761684  0.68117191 -0.8676594
cyl  -0.8521620  1.0000000  0.9020329  0.8324475 -0.69993811  0.7824958
disp -0.8475514  0.9020329  1.0000000  0.7909486 -0.71021393  0.8879799
hp   -0.7761684  0.8324475  0.7909486  1.0000000 -0.44875912  0.6587479
drat  0.6811719 -0.6999381 -0.7102139 -0.4487591  1.00000000 -0.7124406
wt   -0.8676594  0.7824958  0.8879799  0.6587479 -0.71244065  1.0000000
qsec  0.4186840 -0.5912421 -0.4336979 -0.7082234  0.09120476 -0.1747159
vs    0.6640389 -0.8108118 -0.7104159 -0.7230967  0.44027846 -0.5549157
am    0.5998324 -0.5226070 -0.5912270 -0.2432043  0.71271113 -0.6924953
gear  0.4802848 -0.4926866 -0.5555692 -0.1257043  0.69961013 -0.5832870
carb -0.5509251  0.5269883  0.3949769  0.7498125 -0.09078980  0.4276059
     qsec        vs         am          gear       carb       
mpg   0.41868403  0.6640389  0.59983243  0.4802848 -0.55092507
cyl  -0.59124207 -0.8108118 -0.52260705 -0.4926866  0.52698829
disp -0.43369788 -0.7104159 -0.59122704 -0.5555692  0.39497686
hp   -0.70822339 -0.7230967 -0.24320426 -0.1257043  0.74981247
drat  0.09120476  0.4402785  0.71271113  0.6996101 -0.09078980
wt   -0.17471588 -0.5549157 -0.69249526 -0.5832870  0.42760594
qsec  1.00000000  0.7445354 -0.22986086 -0.2126822 -0.65624923
vs    0.74453544  1.0000000  0.16834512  0.2060233 -0.56960714
am   -0.22986086  0.1683451  1.00000000  0.7940588  0.05753435
gear -0.21268223  0.2060233  0.79405876  1.0000000  0.27407284
carb -0.65624923 -0.5696071  0.05753435  0.2740728  1.00000000

* Calculate the correlation between mpg and horsepower


In [21]:
cor(mtcars$mpg, mtcars$hp)

[1] -0.7761684

* Calculate the VIF value

In [24]:
library(car)
vif(cars_reg)

cyl      disp        hp      drat        wt      qsec        vs        am 
15.373833 21.620241  9.832037  3.374620 15.164887  7.527958  4.965873  4.648487 
     gear      carb 
 5.357452  7.908747

Since most of the values are > 5, multicollinearity exists in this model

* Find 3 best subset using AIC

In [27]:
library(olsrr)
ols_step_best_subset(cars_reg)

mindex n  predictors                               rsquare   adjr     
5     1      1 wt                                       0.7528328 0.7445939
14    2      2 cyl wt                                   0.8302274 0.8185189
157   3      3 wt qsec am                               0.8496636 0.8335561
332   4      4 hp wt qsec am                            0.8578510 0.8367919
528   5      5 disp hp wt qsec am                       0.8637377 0.8375334
765   6      6 disp hp drat wt qsec am                  0.8667078 0.8347177
935   7      7 disp hp drat wt qsec am gear             0.8680976 0.8296261
1007  8      8 disp hp drat wt qsec am gear carb        0.8687064 0.8230390
1022  9      9 disp hp drat wt qsec vs am gear carb     0.8689448 0.8153314
1023 10     10 cyl disp hp drat wt qsec vs am gear carb 0.8690158 0.8066423
     predrsq   cp         aic      sbic     sbc      msep     fpe     
5    0.7086954 11.6269926 166.0294 74.37340 170.4266 296.9167 9.857235
14   0.7903761  1.2187315 156.0101 66.19025 161.8730 211.2280 7.210150
157  0.7945881  0.1026357 154.1194 65.71383 161.4481 193.9735 6.801667
332  0.8021094  0.7899838 154.3274 67.22992 163.1218 190.4637 6.854698
528  0.7984327  1.8462076 154.9740 69.30727 165.2341 189.8793 7.007977
765  0.7854830  3.3700162 156.2687 71.92576 167.9946 193.4796 7.317046
935  0.7619427  5.1471984 157.9333 74.80575 171.1250 199.7867 7.735848
1007 0.7316143  7.0496037 159.7853 77.79592 174.4427 207.9040 8.235809
1022 0.7035108  9.0113719 161.7271 80.82774 177.8502 217.4086 8.804149
1023 0.6538246 11.0000000 163.7098 83.87280 181.2986 228.1554 9.437888
     apc       hsp      
5    0.2801228 0.3199103
14   0.2048980 0.2354335
157  0.1932897 0.2239232
332  0.1947968 0.2280149
528  0.1991526 0.2360582
765  0.2079358 0.2501554
935  0.2198373 0.2690730
1007 0.2340452 0.2921795
1022 0.2501962 0.3194249
1023 0.2682058 0.3511772

1. model #3: wt, qsec, am
2. model #2: hp, wt, qsec, am
3. model #3: disp, hp, wt, qsec, am

* using forward selection procedure, find the best model

In [28]:
ols_step_forward_p(cars_reg, pcenter=0.05)


                            Selection Summary                             
-------------------------------------------------------------------------
        Variable                  Adj.                                       
Step    Entered     R-Square    R-Square     C(p)        AIC        RMSE     
-------------------------------------------------------------------------
   1    wt            0.7528      0.7446    11.6270    166.0294    3.0459    
   2    cyl           0.8302      0.8185     1.2187    156.0101    2.5675    
   3    hp            0.8431      0.8263     1.1469    155.4766    2.5115    
-------------------------------------------------------------------------

* repeat using backward elimination method

In [29]:
ols_step_backward_p(cars_reg, pelim=0.05)



                          Elimination Summary                            
------------------------------------------------------------------------
        Variable                  Adj.                                      
Step    Removed     R-Square    R-Square     C(p)       AIC        RMSE     
------------------------------------------------------------------------
   1    cyl           0.8689      0.8153    9.0114    161.7271    2.5900    
   2    vs            0.8687       0.823    7.0496    159.7853    2.5353    
   3    carb          0.8681      0.8296    5.1472    157.9333    2.4877    
   4    gear          0.8667      0.8347    3.3700    156.2687    2.4503    
   5    drat          0.8637      0.8375    1.8462    154.9740    2.4293    
------------------------------------------------------------------------

* fit a MLR where mpg as a response variable and wt, hp, am as covariates also include interaction term hp * wt

In [32]:
model2 = lm(mpg ~ wt + hp + am + wt*hp, data=mtcars)
summary(model2)


Call:
lm(formula = mpg ~ wt + hp + am + wt * hp, data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-2.9845 -1.6580 -0.7407  1.4362  4.5266 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 49.452241   5.280731   9.365 5.69e-10 ***
wt          -8.100558   1.789325  -4.527 0.000108 ***
hp          -0.119303   0.026550  -4.494 0.000119 ***
am           0.125107   1.333431   0.094 0.925942    
wt:hp        0.027488   0.008473   3.244 0.003130 ** 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.192 on 27 degrees of freedom
Multiple R-squared:  0.8848,	Adjusted R-squared:  0.8677 
F-statistic: 51.84 on 4 and 27 DF,  p-value: 2.765e-12


Since the p-value for wt:hp is 0.003 we reject null hypothesis so we do need the interaction term in the model

* Repeat for hp and am

In [33]:
model2 = lm(mpg ~ wt + hp + am + hp*am, data=mtcars)
summary(model2)


Call:
lm(formula = mpg ~ wt + hp + am + hp * am, data = mtcars)

Residuals:
   Min     1Q Median     3Q    Max 
-3.435 -1.510 -0.697  1.284  5.245 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 33.34196    2.79711  11.920 2.89e-12 ***
wt          -3.05617    0.94036  -3.250  0.00309 ** 
hp          -0.02918    0.01449  -2.014  0.05407 .  
am           3.55141    2.35742   1.506  0.14355    
hp:am       -0.01129    0.01466  -0.770  0.44809    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 2.556 on 27 degrees of freedom
Multiple R-squared:  0.8433,	Adjusted R-squared:  0.8201 
F-statistic: 36.33 on 4 and 27 DF,  p-value: 1.68e-10


* This p-value is very large so therefore it's not in the model

* Check if the models are the same

FIGURE OUT WHAT VARIABLES YOU SET TO ZERO SO THAT THE MODELS ARE THE SAME, BUILD THE REDUCED MODELS WITHOUT THOSE VARIABLES, COMBINE IN THE ANOVA TABLE WITH THE FULL MODEL AND CHECK THE F-STATISTIC